sdk/core/Azure.Core.Amqp/src/Shared/AmqpAnnotatedMessageConverter.cs

```diff
@@ -0,0 +1,956 @@
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Amqp.Encoding;
+using Microsoft.Azure.Amqp.Framing;
+
+#nullable enable
+
+namespace Azure.Core.Amqp.Shared
+{
+    /// <summary>
+    ///   Serves as a translator between AmqpAnnotatedMessage and AmqpMessage
+    /// </summary>
+    ///
+    internal static class AmqpAnnotatedMessageConverter
+    {
+        /// <summary>The size, in bytes, to use as a buffer for stream operations.</summary>
+        private const int StreamBufferSizeInBytes = 512;
+
+        /// <summary>The set of mappings from CLR types to AMQP types for property values.</summary>
+        private static readonly IReadOnlyDictionary<Type, AmqpType> AmqpPropertyTypeMap = new Dictionary<Type, AmqpType>
+        {
+            { typeof(byte), AmqpType.Byte },
+            { typeof(sbyte), AmqpType.SByte },
+            { typeof(char), AmqpType.Char },
+            { typeof(short), AmqpType.Int16 },
+            { typeof(ushort), AmqpType.UInt16 },
+            { typeof(int), AmqpType.Int32 },
+            { typeof(uint), AmqpType.UInt32 },
+            { typeof(long), AmqpType.Int64 },
+            { typeof(ulong), AmqpType.UInt64 },
+            { typeof(float), AmqpType.Single },
+            { typeof(double), AmqpType.Double },
+            { typeof(decimal), AmqpType.Decimal },
+            { typeof(bool), AmqpType.Boolean },
+            { typeof(Guid), AmqpType.Guid },
+            { typeof(string), AmqpType.String },
+            { typeof(Uri), AmqpType.Uri },
+            { typeof(DateTime), AmqpType.DateTime },
+            { typeof(DateTimeOffset), AmqpType.DateTimeOffset },
+            { typeof(TimeSpan), AmqpType.TimeSpan },
+        };
+
+        /// <summary>
+        ///   Converts an <see cref="AmqpAnnotatedMessage" /> into an <see cref="AmqpMessage" />.
+        /// </summary>
+        ///
+        /// <param name="sourceMessage">The source message.</param>
+        ///
+        /// <returns>The <see cref="AmqpMessage" /> constructed from the source message.</returns>
+        ///
+        /// <remarks>
+        ///   The caller is assumed to hold ownership over the message once it has been created, including
+        ///   ensuring proper disposal.
+        /// </remarks>
+        ///
+        public static AmqpMessage ToAmqpMessage(AmqpAnnotatedMessage sourceMessage)
+        {
+            var message = sourceMessage switch
+            {
+                _ when sourceMessage.Body.TryGetData(out var dataBody) => AmqpMessage.Create(TranslateDataBody(dataBody!)),
+                _ when sourceMessage.Body.TryGetSequence(out var sequenceBody) => AmqpMessage.Create(TranslateSequenceBody(sequenceBody!)),
+                _ when sourceMessage.Body.TryGetValue(out var valueBody) => AmqpMessage.Create(TranslateValueBody(valueBody!)),
+                _ => AmqpMessage.Create(new Data { Value = new ArraySegment<byte>(Array.Empty<byte>()) })
```

Comment:
Is only one of these guaranteed to be true, or does it not matter which one is selected if so?

Reply:
Yes, only 1 will be true.